File: src/main/java/co/selim/exiftool4j/ExifExtractor.java
package co.selim.exiftool4j;

import java.io.InputStream;

public class ExifExtractor {
    private ExifExtractor() {
    }

    /**
     * Returns exif data that the image contains.
     * This method calls the command line utility 'exiftool' to extract exif data from an image.
     * The input stream will not be closed.
     *
     * @param imageInputStream input stream containing the image
     * @return exif data that the image contains
     */
    public static ExifData extractExifData(InputStream imageInputStream) {
        return ExifExtractorKt.extractExifData(imageInputStream);
    }
}

File: src/main/java/co/selim/exiftool4j/ExifExtractor.kt
package co.selim.exiftool4j

import java.io.InputStream

fun InputStream.extractExifData(): ExifData {
    val exiftoolProcess = ProcessBuilder("exiftool", "-")
        .redirectError(ProcessBuilder.Redirect.INHERIT)
        .start()

    exiftoolProcess.outputStream.use {
        this.copyTo(it)
    }

    val fields = exiftoolProcess.inputStream
        .bufferedReader()
        .useLines(Sequence<String>::toExifMap)

    return ExifData(fields)
}

private fun Sequence<String>.toExifMap(): Map<ExifKey, String> = associateNotNull { line ->
    val key = line.substringBefore(":").trim()
    val value = line.substringAfter(":").trim()
    val exifKey = getExifKey(key)

    if (exifKey != null && exifKey.exclude.none { value.contains(it, true) }) {
        exifKey to value
    } else {
        null
    }
}

private fun <T, K, V> Sequence<T>.associateNotNull(transform: (T) -> Pair<K, V>?): Map<K, V> {
    return mapNotNull(transform).toMap()
}


private fun getExifKey(key: String): ExifKey? = ExifKey.values()
    .firstOrNull { it.names.any { name -> name.equals(key, true) } }
